### PR TITLE
Add group by option to summary.pop_data()

### LINKED ISSUE #112: add group_by option to summary() for class pop_data
summary.pop_data()

### PR DESCRIPTION
**Pull request by chrisorwa**
Provide a `strata` argument to the `summary.pop_data()` function to allow result stratification.

### CODE REVIEW

File: R/summary.pop_data.R [hunk 1]
```diff
@@ -12,21 +13,40 @@
...
   cat("\nDistribution of age: \n\n")
   age_summary <-
     ages %>%
-    pull("age") %>%
-    summary() %>%
+    select(
+      age_column,
+      all_of(strata)
+    ) %>%
+    group_by(across(all_of(strata))) %>%
```
Comment (d-morrison) [inline R/summary.pop_data.R]: please avoid using `group_by()` unless the grouping needs to persist across multiple steps; otherwise use the `.by` argument in `summarise()`, `mutate()`, etc.

File: R/summary.pop_data.R [hunk 2]
```diff
@@ -12,21 +13,40 @@
 #' library(dplyr)
 #'
 #' xs_data <- load_pop_data("https://osf.io/download//n6cp3/")
+#' summary(xs_data, strata = "Country")
 #'
-#' summary(xs_data)
-#'
-summary.pop_data <- function(object, ...) {
+summary.pop_data <- function(object, strata = "Country", ...) {
```
Comment (d-morrison) [inline R/summary.pop_data.R]: I'm not sure we should default to stratifying by "Country"; I'd think the default should be `NULL`? @kaiemjoy  please advise

**Review by d-morrison — CHANGES_REQUESTED** [review]
Looks good overall; see comments.

File: R/summary.pop_data.R (hunk 1 above)
Comment (chrisorwa) [reply R/summary.pop_data.R]: I have made the change.

File: R/summary.pop_data.R (hunk 2 above)
Comment (chrisorwa) [reply R/summary.pop_data.R]: I have set strata as `NULL` and have the old summary when null and provide stratification by the provided `strata`

File: R/summary.pop_data.R [hunk 3]
```diff
@@ -12,44 +13,130 @@
...
+
+    age_summary <-
+      ages %>%
+      pull(age_column) %>%
+      summary() %>%
+      print()
+
+    cat("\nn =", nrow(ages), "\n")
+
+    cat("\nDistribution of age: \n\n")
```
Comment (d-morrison) [inline R/summary.pop_data.R]: why did the order of these commands change? what does the output look like? Please provide before-and-after screenshots.

File: R/summary.pop_data.R [hunk 4]
```diff
@@ -12,44 +13,130 @@
...
+        .by = strata
+      ) %>%
+      print()
+
+    cat("\nDistributions of antigen-isotype measurements:\n\n")
+
+    ab_summary <-
+      object %>%
+      dplyr::summarize(
+        .by = .data$antigen_iso,
```
Comment (d-morrison) [inline R/summary.pop_data.R]: don't we need to add `strata` to the `.by` argument here?

File: R/summary.pop_data.R [hunk 5]
```diff
@@ -12,44 +13,130 @@
...
+      as.data.frame() %>%
+      print()
+
+    to_return <- list(
+      n = nrow(ages),
+      age_summary = age_summary,
+      ab_summary = ab_summary
+    )
+
+    return(invisible(to_return))
```
Comment (d-morrison) [inline R/summary.pop_data.R]: Is this code identical in both the "if" and "else" clauses? if so, please don't duplicate it; move it after the conditional lines.

File: R/summary.pop_data.R [hunk 6]
```diff
@@ -12,44 +13,130 @@
...
+  }
```
Comment (d-morrison) [inline R/summary.pop_data.R]: do we need separate code for stratified and unstratified cases? It looks like the stratified code can already handle `strata = NULL`. Please see if you can simplify this code; we want to avoid copy-pasting whenever possible.

**Review by d-morrison — CHANGES_REQUESTED** [review]
Coming along; please see comments, and please add some unit testing for `summary.pop_data()`

File: R/summary.pop_data.R (hunk 6 above)
Comment (chrisorwa) [reply R/summary.pop_data.R]: I have simplified the code but we still need the stratified and unstratified case. Unstratified is the older code version, and the new version extends it to have the stratification.

File: R/summary.pop_data.R (hunk 5 above)
Comment (chrisorwa) [reply R/summary.pop_data.R]: I have made this change.

File: R/summary.pop_data.R (hunk 4 above)
Comment (chrisorwa) [reply R/summary.pop_data.R]: Yes we do.

File: R/summary.pop_data.R (hunk 3 above)
Comment (chrisorwa) [reply R/summary.pop_data.R]: The commands need not change. I have switched them back to the original place.

File: R/summary.pop_data.R [hunk 7]
```diff
@@ -21,122 +21,67 @@ summary.pop_data <- function(object, strata = NULL, ...) {
...
-        age_mean = mean(.data[[age_column]]),
-        age_third_quartile = quantile(.data[[age_column]], 0.75),
-        age_max = max(.data[[age_column]]),
-        .by = strata
-      ) %>%
-      print()
+  cat("\nn =", nrow(ages), "\n")
 
-    cat("\nDistributions of antigen-isotype measurements:\n\n")
+  cat("\nDistribution of age: \n\n")
```
Comment (d-morrison) [inline R/summary.pop_data.R]: lines 49-51 are still out of order relative to lines 33-47, causing the message "Distribution of age:" to appear after the corresponding table:

image


The message is supposed to appear before the first table.

Comment (d-morrison) [reply R/summary.pop_data.R]: also the prefix "age_" in the first table is redundant; the whole table is about age. please remove the prefix from that table.

File: R/summary.pop_data.R [hunk 8]
```diff
@@ -12,35 +13,66 @@
...
+      age_first_quartile = quantile(.data[[age_column]], 0.25),
+      age_median = median(.data[[age_column]]),
+      age_mean = mean(.data[[age_column]]),
+      age_third_quartile = quantile(.data[[age_column]], 0.75),
+      age_max = max(.data[[age_column]]),
+      .by = strata
+    ) %>%
     print()
 
-  cat("\nDistributions of antigen-isotype measurements:\n\n")
```
Comment (d-morrison) [inline R/summary.pop_data.R:32]: why was this line deleted?

File: tests/testthat/test-summary.pop_data.R [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    summary(strata = "Country") %>%
+    magrittr::extract2("age_summary") %>%
+    pull(age_first_quartile) %>%
+    mean()
+
+  expect_equal(
+    unstratified_summary %>% pull(age_first_quartile) %>% as.numeric(),
+    first_quartile_country,
+    tolerance = 0.5
+  )
```
Comment (d-morrison) [inline tests/testthat/test-summary.pop_data.R]: as discussed in teams, we want to test:

does the function execute without errors for standard input values (`strata = NULL`, strata = "Country", etc)
does the function return what we expect it to for various input values? (pre-save some example outputs as internal data, and check for equality of current outputs)

we don't expect stratified and unstratified results to be similar to each other.

**Review by d-morrison — CHANGES_REQUESTED** [review]
coming along; please see notes.

File: R/summary.pop_data.R (hunk 7 above)
Comment (chrisorwa) [reply R/summary.pop_data.R]: This changes have been made.

File: tests/testthat/test-summary.pop_data.R (hunk 9 above)
Comment (chrisorwa) [reply tests/testthat/test-summary.pop_data.R]: I have written new tests to replace these.

**Comment by chrisorwa** [comment]
@d-morrison I have renamed the `nlm_exit_codes` file to `data_objects`  and included other objects that should be on `R/sysdata.rda` .  The reason is to append all the objects appended to `R/sysdata.rda` - running separate scripts and appending the results is impossible. 

On the `test-summary.pop_data`, I have suppressed the warnings since it seems it not related to the functions - I rewrote the function to escape use of select() yet the warning persisted.

File: R/add_point_to_graph.R [hunk 10]
```diff
@@ -5,7 +5,7 @@ add_point_to_graph <- function(
     point_data =
       tibble(
         x = fit$estimate %>% exp(),
-        y = llik(.data$x, ...),
+        y = log_likelihood(.data$x, ...),
```
Comment (d-morrison) [inline R/add_point_to_graph.R]: please move this rename to a separate PR

File: R/summary.pop_data.R [hunk 11]
```diff
@@ -22,38 +22,35 @@ summary.pop_data <- function(object, strata = NULL, ...) {
   id_column <- object %>% get_id_var()
 
   # create a list of the columns
-  cols <- c(age_column, id_column, strata)
```
Comment (d-morrison) [inline R/summary.pop_data.R]: why was `id_column` removed? Please show us output before and after this change

File: data-raw/data_objects.R [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    start = start
+  ) %>%
+  mutate(
+    ageCat = NULL,
+    antigen.iso = paste(collapse = "+", "HlyE_IgG")
+  ) %>%
+  structure(noise.parameters = noise)
+
+
+usethis::use_data(typhoid_results, summary_country, nlm_exit_codes, overwrite = TRUE, internal = TRUE)
```
Comment (d-morrison) [inline data-raw/data_objects.R]: the `typhoid_results` and `summary_country` objects are only used for testing, correct? if so, they don't belong in sysdata, which gets installed. (I think I told you to put them in `sysdata`, but that was incorrect.) 

These objects belong under `tests/testthat/fixtures`; see <https://r-pkgs.org/testing-design.html#storing-test-data>. (and the code to create them should be kept separate from the code for `sysdata`).

File: data-raw/nlm_exit_codes.R
Comment (d-morrison) [inline data-raw/nlm_exit_codes.R:1]: On further thought, I think we can get rid of `sysdata` altogether, and just leave `nlm_exit_codes` as an unexported object defined in this file; see
<https://r-pkgs.org/data.html#sec-data-sysdata:~:text=Sometimes%20the%20objects%20you%20need%20are%20small%20and%20simple%20enough%20that%20you%20can%20define%20them%20with%20c()%20or%20data.frame()%20in%20the%20code%20below%20R/%2C%20perhaps%20in%20R/data.R>

File: data-raw/typhoid_results.qmd [hunk 13]
```diff
@@ -92,7 +73,10 @@ typhoid_results <- fit %>%
     ageCat = NULL,
     antigen.iso = paste(collapse = "+", "HlyE_IgG")
   ) %>%
-  structure(noise.parameters = cond.hlye.IgG)
+  structure(noise.parameters = noise)
 
 usethis::use_data(typhoid_results, internal = TRUE, overwrite = TRUE)
```
Comment (d-morrison) [inline data-raw/typhoid_results.qmd]: see above; we now want to store this data in the testing subdirectories as described in <https://r-pkgs.org/testing-design.html#storing-test-data>

File: tests/testthat/test-est.incidence.R [hunk 14]
```diff
@@ -1,66 +1,46 @@
...
 
     # compare with `typhoid_results` from data-raw/typhoid_results.qmd
-
-    expect_equal(
-      object = fit,
-      expected = typhoid_results
-    )
+      expect_equal(
+        object = fit,
+        expected = typhoid_results
```
Comment (d-morrison) [inline tests/testthat/test-est.incidence.R]: you'll need to explicitly load `typhoid_results` after making the changes previously mentioned

File: tests/testthat/test-summary.pop_data.R [hunk 15]
```diff
@@ -6,87 +6,33 @@ xs_data <- load_pop_data(
...
-    mean()
-
-  expect_equal(
-    unstratified_summary %>% pull(age_first_quartile) %>% as.numeric(),
-    first_quartile_country,
-    tolerance = 0.5
+test_that("`summary.pop_data()` produces an error when wrong stratification is provided", {
+  expect_error(
+    object = xs_data %>% summary(strata = "province"),
+    regexp = "Element `province` doesn't exist.",
```
Comment (d-morrison) [inline tests/testthat/test-summary.pop_data.R:12]: "Element" should be "Column"

File: tests/testthat/test-summary.pop_data.R [hunk 16]
```diff
@@ -6,87 +6,33 @@ xs_data <- load_pop_data(
...
-    pull(age_median) %>%
-    mean()
-
-  expect_equal(
-    unstratified_summary %>% pull(age_median),
-    median_country,
-    tolerance = 0.5
-  )
+test_that("`summary.pop_data()` does not produce an error when NULL", {
+  # suppress warning avoids a deprecation tidyverse warning on use of select(data)
```
Comment (d-morrison) [inline tests/testthat/test-summary.pop_data.R]: what warning? and why do we need to suppress it?

**Review by d-morrison — CHANGES_REQUESTED** [review]
Making progress; see comments below.

File: tests/testthat/test-summary.pop_data.R (hunk 15 above)
Comment (chrisorwa) [reply tests/testthat/test-summary.pop_data.R:12]: It should be "Element" because we are using a regex (`regexp`) to capture the error message (see below) below. It has to be the exact error message thrown otherwise we can't catch the error. 
image

File: data-raw/typhoid_results.qmd (hunk 13 above)
Comment (chrisorwa) [reply data-raw/typhoid_results.qmd]: It seems this link is not working.

File: R/add_point_to_graph.R (hunk 10 above)
Comment (chrisorwa) [reply R/add_point_to_graph.R]: This is well-noted.

File: data-raw/nlm_exit_codes.R [hunk 17]
```diff
@@ -0,0 +1,9 @@
+nlm_exit_codes <- c(
+  "1" = "1: relative gradient is close to zero, current iterate is probably solution.",
+  "2" = "2: successive iterates within tolerance, current iterate is probably solution.",
+  "3" = "3: Last global step failed to locate a point lower than x. Either x is an approximate local minimum of the function, the function is too non-linear for this algorithm, or `stepmin` in `est.incidence()` (a.k.a. `steptol` in `nlm()`) is too large.",
+  "4" = "4: iteration limit exceeded; increase `iterlim`.",
+  "5" = "5: maximum step size `stepmax` exceeded five consecutive times. Either the function is unbounded below, becomes asymptotic to a finite value from above in some direction or `stepmax` is too small."
+)
+
+usethis::use_data(nlm_exit_codes, overwrite = TRUE, internal = TRUE)
```
Comment (d-morrison) [inline data-raw/nlm_exit_codes.R]: is line 9 necessary?

File: tests/testthat/test-est.incidence-status.R [hunk 18]
```diff
@@ -1,35 +1,35 @@
...
-  est_false <- est.incidence(
-    pop_data = xs_data_false %>% filter(Country == "Pakistan"),
-    curve_params = curves,
-    noise_params = noise %>% filter(Country == "Pakistan"),
-    antigen_isos = c("HlyE_IgG", "HlyE_IgA")
-  )
-
-  expect_equal(est_true, est_false)
-})
+# test_that("`est.incidence()` produces expected results", {
```
Comment (d-morrison) [inline tests/testthat/test-est.incidence-status.R]: why is this code commented out? can we just delete it if it's not needed?

File: data-raw/typhoid_results.qmd (hunk 13 above)
Comment (d-morrison) [reply data-raw/typhoid_results.qmd]: try again? works for me.

File: data-raw/typhoid_results.R
Comment (d-morrison) [inline data-raw/typhoid_results.R:1]: please replace all manual fixtures with automated snapshots: 
https://r-pkgs.org/testing-basics.html#sec-snapshot-tests

**Review by d-morrison — CHANGES_REQUESTED** [review]
Thanks for this work! Almost there. Please:

- [x]  address the comments below
- [x]  review and incorporate this patch: https://github.com/UCD-SERG/serocalculator/pull/171

File: data-raw/nlm_exit_codes.R (hunk 17 above)
Comment (chrisorwa) [reply data-raw/nlm_exit_codes.R]: No, it is not necessary. The code builds without it.

File: data-raw/typhoid_results.R
Comment (chrisorwa) [reply data-raw/typhoid_results.R:1]: This is done.

File: tests/testthat/test-est.incidence.R [hunk 19]
```diff
@@ -37,16 +37,45 @@ test_that(
...
     # load the RDS file
-    typhoid_results <- readRDS(typhoid_results_path)
+    typhoid_results <- readRDS(test_path("fixtures", "typhoid_results.rds"))
 
-    # compare with `typhoid_results` from data-raw/typhoid_results.qmd
-      expect_equal(
-        object = fit,
-        expected = typhoid_results
-      )
+    expect_snapshot(x = typhoid_results)
```
Comment (d-morrison) [inline tests/testthat/test-est.incidence.R]: Please re-read https://r-pkgs.org/testing-basics.html#sec-snapshot-tests more carefully; this is an incorrect use of `expect_snapshot()`.

In particular: Please remove *all* manually created fixture objects. The input to `expect_snapshot()` should be the result of running the relevant function (in this case, `est.incidence()`) during the test; you can't just load a pre-saved object and snapshot-test it. 

Please follow the example I provided in [my previous patch](https://github.com/UCD-SERG/serocalculator/pull/171).

File: R/sysdata.rda
Comment (d-morrison) [inline R/sysdata.rda:1]: I think we can delete this file now; please try removing it (and re-check and re-test).

File: tests/testthat/fixtures/typhoid_results.rds
Comment (d-morrison) [inline tests/testthat/fixtures/typhoid_results.rds:1]: please delete this file

File: data-raw/typhoid_results.qmd
Comment (d-morrison) [inline data-raw/typhoid_results.qmd:1]: This file can also be deleted after moving its contents into a snapshot test.

**Review by d-morrison — CHANGES_REQUESTED** [review]
Almost there! Please see comments below.

File: tests/testthat/test-est.incidence.R (hunk 19 above)
Comment (chrisorwa) [reply tests/testthat/test-est.incidence.R]: This has been corrected.

File: R/sysdata.rda
Comment (chrisorwa) [reply R/sysdata.rda:1]: I removed `sysdata.rda` and encountered an error indicating that `nlm_exit_codes` was not found. When I re-added the code `usethis::use_data(nlm_exit_codes, overwrite = TRUE, internal = TRUE)` to recreate `sysdata.rda` on the `nlm_exit_codes.R`, the build was successful.

File: tests/testthat/fixtures/typhoid_results.rds
Comment (chrisorwa) [reply tests/testthat/fixtures/typhoid_results.rds:1]: This is done.

File: R/sysdata.rda
Comment (d-morrison) [reply R/sysdata.rda:1]: @chrisorwa please screenshot the error; also, please try moving `nlm_exit_codes.R` from `data-raw/` to `R/`, as specified in https://r-pkgs.org/data.html#sec-data-sysdata:

> Sometimes the objects you need are small and simple enough that you can define them with [c()](https://rdrr.io/r/base/c.html) or [data.frame()](https://rdrr.io/r/base/data.frame.html) **in the code below R/**, perhaps in R/data.R

Comment (d-morrison) [reply R/sysdata.rda:1]: @chrisorwa here's a test patch: https://github.com/UCD-SERG/serocalculator/pull/180 it passed checks on my local; let's see if it passes the online checks.

Comment (chrisorwa) [reply R/sysdata.rda:1]: image

Comment (chrisorwa) [reply R/sysdata.rda:1]: @d-morrison the patch passed the online checks.

**Review by d-morrison — APPROVED** [review]
Looks great - thanks for all your hard work on this PR!

**Comment by chrisorwa** [comment]
Cheers!!!
